Paper: Any notebook served: authoring and sharing reusable interactive widgets

nvictus opened the pull request:
If you are creating this PR in order to submit a draft of your paper, please name your PR with `Paper: <title>`. An editor will then add a `paper` label and GitHub Actions will be run to check and build your paper.

See the [project readme](https://github.com/scipy-conference/scipy_proceedings#preview-your-paper) for more information.

_Editor_: <!--editor--> Sanhita Joshi @sanhitamj <!--end-editor-->

_Reviewers_: <!--reviewers-list--> 
- Rashmi K Yamakanamardi @rasshmi99
- Veronica gathoni @chegeveronica 
<!--end-reviewers-list-->

nvictus committed:
anywidget initial commit

nvictus committed:
anywidget updates

nvictus committed:
Update manuscript

manzt committed:
Rewrite last paragraph

manzt committed:
Update acknowledgements

manzt committed:
Merge pull request #2 from nvictus/manzt/edits

Expand on use cases

manzt committed:
More edits (#3)

* Update bib

* edits

manzt committed:
more edits

manzt committed:
update figures

manzt committed:
new colors

manzt committed:
touch up abstract

manzt committed:
add supernova bibtex

manzt committed:
add cev paper

manzt committed:
e.g.,

manzt committed:
Remove trailing )

nvictus committed:
moar edits

nvictus committed:
Merge branch '2024' of https://github.com/nvictus/scipy_proceedings into 2024

manzt committed:
Fix dangling this

manzt committed:
Remove passive voice

manzt committed:
add fowler di paper

manzt committed:
more trevor edits

nvictus committed:
moare edots

manzt committed:
Add AFM description

manzt committed:
Add ending

manzt committed:
Fix myst id

nvictus committed:
moor eduts

nvictus committed:
Merge branch '2024' of https://github.com/nvictus/scipy_proceedings into 2024

manzt committed:
fold text

nvictus committed:
keywords

nvictus committed:
Merge branch '2024' of https://github.com/nvictus/scipy_proceedings into 2024

nvictus committed:
muur edats

manzt commented on papers/manz_anywidget/main.md:
```diff
@@ -0,0 +1,336 @@
+Widgets are intended to be pluggable components, similar to kernels. However,
+only the protocol for communication between kernel and front-end widget code,
+known as the [Jupyter Widgets Message
+Protocol](https://github.com/jupyter-widgets/ipywidgets/blob/main/packages/schema/messages.md),
```
We probably want a specific commit hash for this URL.

manzt replied:
```suggestion
Protocol](https://github.com/jupyter-widgets/ipywidgets/blob/c65dc47/packages/schema/messages.md),
```

manzt committed:
Update papers/manz_anywidget/main.md

manzt committed:
fix: links

manzt committed:
Add DOIs

manzt committed:
fix roles

manzt commented on papers/manz_anywidget/myst.yml:
```diff
@@ -0,0 +1,72 @@
+      orcid: 0000-0003-0327-8297
+      affiliations:
+        - Department of Biomedical Informatics, Harvard Medical School, Boston, MA, USA
+      roles:
```
```suggestion
     fix-me:
```

manzt committed:
Update refs

manzt committed:
Add support for other JCPs

manzt committed:
fix refs

manzt committed:
Add pngs

manzt committed:
Update ecma citation

nvictus committed:
Update metadata

nvictus committed:
Cite bridges and add community acknowledgement

manzt committed:
Fix curvenote checks

manzt committed:
chore: Update anywidget ref with DOI

chegeveronica commented:
Hi. I'm Veronica Chege, a member of the review team.

This is a very well-written paper. I love how the project offers a practical solution to the challenges of authoring and distributing reusable widgets. The Anywidget project aims to simplify the authoring, distribution, and execution of interactive web-based widgets across various Jupyter-compatible platforms. The authors describe their approach's motivation, methodology, and potential impact, highlighting the improvements in reusability and interoperability of widgets. As my initial review, I'd suggest adding Anywidget's performance metrics compared to the other tools and user feedback to strengthen the argument for its effectiveness. Generally, this is great to work from the authors.

manzt committed:
update manuscript with reviewer feedback

nvictus commented:
Thank you for the very positive review @chegeveronica! 

While we understand the value of quantitative metrics, it is challenging to imagine a direct comparison of anywidget to other tools since there is no other tool that accomplishes what anywidget does. A widget authored with anywidget has the same runtime performance characteristics to widgets created using other methods. We have added clarifying language about this in the manuscript #[521a03a](https://github.com/scipy-conference/scipy_proceedings/pull/930/commits/521a03a6b9b3bc06c0de7cc3fcd4e5ef8ea0b1c1). 

The difference is that anywidget eliminates sources of development friction and introduces conveniences for developers. Short of performing a controlled study of developer velocity, we think anywidget’s adoption curve itself is a testament to the problems that it solves for individual developers. As of 2024, not a single traditional Jupyter widget has been released (**Figure 5**) and 22 traditional widgets have been ported to anywidget. The number of new Jupyter widgets in the ecosystem at the time of writing (mid-year) already exceeds the peak year of widget creation (2020, during the pandemic). In fact, Figure 5 needs to be updated because 5 additional anywidgets have been published since this manuscript was submitted.

To our knowledge, the only other way to author custom Jupyter Widgets is via cookiecutters or repository templates, which require the use of front-end tooling, build configurations, and fragmented packaging steps. The “Hello, World” example from the custom Jupyter Widgets cookiecutter requires a project template with 53 other files (**~1500 LOC**), a large surface area. The identical widget with anywidget can be created in less than 20 lines of code, entirely from within a notebook cell. We have added a reference in the manuscript linking to the [section of our documentation](https://anywidget.dev/blog/introducing-anywidget/#a-solution-with-crumbs) that compares the two implementations.

mwcraig commented on papers/manz_anywidget/main.md:
```diff
@@ -0,0 +1,368 @@
+objects residing in the kernel [@doc_juparch]. For instance, the
+[ipywidgets](https://github.com/jupyter-widgets/ipywidgets) library provides
+basic form elements like buttons, sliders, and dropdowns to adjust individual
+variables. Other community projects offer interactive visualizations for
```
The core package also defines the communication protocol for widgets between the front-end and the kernel.

mwcraig commented on papers/manz_anywidget/main.md:
```diff
@@ -0,0 +1,368 @@
+understand both packaging ecosystems. Moreover, this divided system is
+incompatible with shared, multi-user environments like
+[JupyterHub](https://jupyter.org/hub), where users can only install kernel-side
+widget code in their custom kernels. Since front-end widget code must be
```
This seems like an over-broad statement. Do you mean for it to apply only to custom kernels or to JupyterHub in general?

manzt replied:
I have clarified this section to avoid the use of "custom kernels." The issue is that in shared Jupyter envs (like JupyterHub) users have the ability to install regular Python packages into their own environments. Widgets, in contrast, require installation in the _shared_ webserver, typically by an admin. This means that users generally do not have the freedom to independently install and use various widgets.

mwcraig commented on papers/manz_anywidget/main.md:
```diff
@@ -0,0 +1,368 @@
+  environments with custom visualizations and interactive elements that
+  communicate directly with user code and data. While this bidirectional
+  communication makes the widget system powerful, its architecture is currently
+  tightly coupled to platforms. As a result, widgets are complex and error-prone
```
`widgets are complex and error-prone...` -- I think this needs some qualification. It is complex and error-prone if it is a custom widget that includes front-end components. It is not complex if the widget is Python-only.

manzt replied:
Thanks for the comment. We have added an admonition clarifying that we are referring to custom widgets with front-end components.

mwcraig commented on papers/manz_anywidget/main.md:
```diff
@@ -0,0 +1,368 @@
+or does not match with the user’s kernel-side widget code.
+
+:::{figure} dev-before-afm-01.svg
+:label: fig:before-afm
```
A question about this figure: the color of the blue puzzle piece for "widget c" is a different shade of blue than the rest of the blue pieces. Is that intended to be meaningful?

manzt replied:
Thanks for the feedback. We have updated the figure caption explaining these nuances.

mwcraig commented on papers/manz_anywidget/main.md:
```diff
@@ -0,0 +1,368 @@
+> both in the webserver and the kernel. The kernel components are easy to
+> install, because you can rely on the language’s built-in tools. The static
+> assets for the webserver complicate things, because an extra step is required
+> to let the webserver know where the assets are.
```
This is a solved problem for the notebook/lab environment though -- the cookie cutter provides the template to get an easily installable package for JupyterLab. That is not to say that the system is not complicated...I've messed it up multiple times!

manzt replied:
We acknowledge that this issue is generally solved for single-user Jupyter and JupyterLab. However, it still remains a challenge for shared Jupyter environments (like JupyterHub) where users do not have the ability to install extensions (and therefore traditional widgets).

mwcraig commented on papers/manz_anywidget/main.md:
````diff
@@ -0,0 +1,368 @@
+```
+An anywidget front-end module (AFM) with initialization and rendering lifecycle
+methods. For familiarity, AFM methods use naming conventions from traditional
+Jupyter Widgets; however, AFM narrows down the APIs. _Methods_: The
````
`AFM narrows down the APIs` -- can you elaborate on this a bit in the text?

manzt replied:
AFM narrows down the APIs by defining a subset of interfaces derived from traditional Jupyter widgets. It intentionally strips off many interfaces and methods, retaining only those "essential." This narrowing allows new environments to load and execute anywidgets, which is not possible with traditional Jupyter widgets due to their more complex and coupled interfaces. Implementing the full set of traditional Jupyter widget front-end APIs in new environments would be very challenging. We've linked a blog post that explains this concept in detail.

mwcraig commented on papers/manz_anywidget/main.md:
```diff
@@ -0,0 +1,368 @@
+models for state or UI. It allows widgets to be defined without dependencies,
+reducing boilerplate and preventing lock-in. Authors may still use third-party
+JavaScript libraries or tooling, but these are no longer necessary for JCP
+compatibility, publishing, or user installation.
```
A little more discussion here of any limitations might be helpful.

manzt replied:
Thanks for the feedback! We have added a limitations section .

mwcraig commented on papers/manz_anywidget/main.md:
```diff
@@ -0,0 +1,368 @@
+standalone scripts or packages – just like kernel-side programs
+[@fig:after-afm]. In contrast, developing traditional Jupyter Widgets is a
+cumbersome process limited to the Jupyter Notebook and JupyterLab platforms. It
+involves using a project generator [@js_cookiecutter; @ts_cookiecutter] to
```
The JavaScript cookiecutter now directs people to anywidget 😀 so maybe drop it here?

manzt replied:
Great suggestion. We've added that the JS cookiecutter now directs folks to anywidget.

mwcraig commented on papers/manz_anywidget/main.md:
```diff
@@ -0,0 +1,368 @@
+Serving AFMs and other static assets from the kernel
+removes the need to publish widget kernel-side and front-end code separately and
+coordinate their releases. For example, many JCPs retrieve traditional widget
+Javascript code from [npm](https://www.npmjs.com/), misusing the registry for
```
More detail about how this constitutes mis-use of the registry would be helpful.

manzt replied:
Thanks for the question. Jupyter Widgets are dual Python / JavaScript programs. JCPs typically retrieve widget JavaScript from Content Delivery Networks (CDNs) like jsDelivr or unpkg, which mirror versioned contents published to npm. This approach may not align well with npm's primary purpose of hosting reusable JavaScript packages, meant to be installed in other _JavaScript_ projects via npm's client. 

Widget front-end code is specialized and not really independently reusable outside of JCPs (and the Python component). Using npm this way effectively makes it a host for static files, which may not fully align with npm's [acceptable content policy](https://docs.npmjs.com/policies/open-source-terms#acceptable-content) that emphasizes functional compatibility with the npm command-line client.

mwcraig commented on papers/manz_anywidget/main.md:
```diff
@@ -0,0 +1,368 @@
+distributing specialized programs rather than reusable JavaScript modules.
+Instead, with anywidget, developers can publish a widget (kernel-side module,
+AFM, and stylesheets) as a unified package to the distribution channels
+relevant to the kernel language, such as the [Python Package Index](https://pypi.org/).
```
To be clear, a widget that uses the TypeScript cookiecutter can also be distributed via PyPI, though it might only work in JupyterLab and not other JCPs.

manzt replied:
This is a good point, thanks for making the clarification. You're right that widgets using the TypeScript cookiecutter can be distributed via PyPI only, though they may only work in JupyterLab. However, this approach has limitations. It doesn't work across all JCPs and breaks other ipywidgets features like embedding widgets in standalone HTML (e.g.,`ipywidgets.embed.embed_minimal_html`). We consider publishing front-end code separately as crucial to meet most users' expectations for custom Jupyter Widgets.

mwcraig commented on papers/manz_anywidget/main.md:
```diff
@@ -0,0 +1,368 @@
+compatible with more platforms, but achieving compatibility involves trade-offs
+[@Wang2024]. The full capabilities of the widget system, such as
+bidirectional communication, are often inaccessible to authors due to
+development difficulty and maintenance efforts. Adopting standards can minimize
```
Do you mean specifically standards for distribution?

manzt replied:
Thanks for the question. We're referring to standardizing how front-end code is authored, not just distributed. Specifically, we mean standardizing (i.e., AFM) so widget code isn't tailored to specific Jupyter environments. This benefits authors by reducing cross-platform development effort and increasing familiarity through consistent patterns across projects.

mwcraig commented on papers/manz_anywidget/main.md:
```diff
@@ -0,0 +1,368 @@
+
+:::{figure} widgetstats.png
+:align: center
+:label: fig:widgetstats
```
I think there needs to be more detail about how this was constructed.

manzt replied:
Thanks for pushing us to improve both our discovery process and in-turn the widgets dataset!

mwcraig commented on papers/manz_anywidget/main.md:
```diff
@@ -0,0 +1,368 @@
+order to support live development of custom widgets without any front-end
+tooling. For example, adjusting a widget's appearance, such as a chart's color
+scheme, updates the view instantly without re-executing cells or refreshing the
+page.
```
This is a really nice feature!

manzt replied:
Thanks!

mwcraig reviewed (commented):
Hi 👋, apologies for the delay in writing this review 😬. 

I thought it would be helpful to preface the review by saying I'm a member of the Jupyter Widgets Council; I think I am able to do a fair review of the paper and that this doesn't pose a conflict of interest. I'm stating it up front, though, in case anyone involved feels otherwise. If so, I'm happy to step back as a reviewer.

I was also an enthusiastic attendee of your tutorial at SciPy 2024, and wrote a small package for fun during the sprints called `soupernova` that is included in your gallery/list of anywidget packages.

Some high level comments:

1. The paper is very clear and well written. I have a couple of minor comments on the text/figures as inline comments.
2. How should the citations appear in the render paper? When I read the preview most of the citations should up as gold-ish text that didn't link to anything. Not sure if that is expected or not.
1. I think it would be useful to, at some point early in the paper, specify that the challenging part of writing a custom widget applies only to widgets that have front-end components. Widget packages that are python-only do not have that restriction, and can be useful.
3. I think there needs to be more detail about how you are calculating the number of widget packages that are ipywidgets vs anywidget vs converted. I tried reading the code in `search.ts` in [anywidget-usage](https://github.com/manzt/anywidget-usage/tree/main) to understand how the search is done but I don't. When I ran what I thought was the query for ipywidget-based packages in the web interface for github code search I didn't get very many hits. There is no need to explain line-by-line how that code works, but prose explaining how you searched for packages and arrived at your numbers would be helpful in understanding the impact of anywidget. I found a number of ipywidget-based widgets that include TypeScript or JavaScript components which are not included, as far as I can tell, in your counts -- this is not meant to be a comprehensive list, it is simply meant to indicate that it seems some things are being missed:
    1. https://github.com/tensorware/ipypandas
    4. https://github.com/davidbrochart/ipyspin
    5. https://github.com/jupyter-robotics/ipynao
    6. https://github.com/mwcraig/ipyevents (disclaimer: this one is mine)
    7. https://github.com/jupyter-widgets/jupyterlab-sidecar
    8. ...
9. The previous point doesn't take away from your main conclusion that the anywidget approach is easier than the cookie-cutter based approach -- it is definitely less complicated 😀! I also don't doubt that anywidget has increased the number of custom widgets being written that contain javascript.
10. I think there should be some acknowledgement of any downsides of the anywidget approach. The main one I see in my limited experience with anywidget is that importing javascript modules from URLs at runtime means they cannot be used offline.

fwkoch commented on papers/manz_anywidget/mybib.bib:
```suggestion
  url={https://anywidget.dev/blog/introducing-anywidget/#a-solution-with-crumbs}
```

This typo is breaking all the references...

manzt committed:
Update papers/manz_anywidget/mybib.bib

Co-authored-by: Franklin Koch <franklin@curvenote.com>

rowanc1 commented:
I am leaving an enthusiastic review of the paper, especially based on many of the conversations on integrating with MyST Markdown for scientific articles (like this one!). One of the things I would love to see better articulated is that widgets can work without the kernel (in many cases, e.g. educational widgets or static data set explorers).

This opens up a lot of possibilities for reuse of these modular widgets beyond jupyter compatible ecosystems, but are also compatible and can be upgraded by a kernel, when available. This opens possibilities when combined with tools like MyST, which may or may not have a kernel available.

If you can't figure out how to make the references work, please let me or @fwkoch know.

fwkoch commented on papers/manz_anywidget/mybib.bib:
```suggestion
  url={https://anywidget.dev/blog/introducing-anywidget/#a-solution-with-crumbs},
```

Oh man, there's more... 😅 At least there are nice errors coming through the build log:

![image](https://github.com/user-attachments/assets/01d890f4-4469-482e-a1e0-5a6f826696be)

fwkoch replied:
Ok, better now!

![image](https://github.com/user-attachments/assets/eca232cd-8ea7-41d8-8ecc-054ba4921d77)

fwkoch committed:
Update papers/manz_anywidget/mybib.bib

mwcraig commented:
> 4\. I think there needs to be more detail about how you are calculating the number of widget packages that are ipywidgets vs anywidget vs converted. I tried reading the code in `search.ts` in [anywidget-usage](https://github.com/manzt/anywidget-usage/tree/main) to understand how the search is done but I don't. When I ran what I thought was the query for ipywidget-based packages in the web interface for github code search I didn't get very many hits.

I realized this morning that my intent in bringing this up is not that you will find every single widget package there is -- it is just to clearly explain what search criteria you used. No matter what you do you will almost certainly miss some.

sanhitamj commented:
@chegeveronica and @mwcraig thank you so much for a detailed review.

@nvictus here are some more reviewer comments to address.

sanhitamj commented:
Hi @nvictus , 2nd Sept is the deadline to address the reviewer comments.

manzt committed:
Add admonition

manzt committed:
Clarify multi-user environment issue

manzt committed:
Clarify puzzle pieces in figure caption

manzt committed:
Clarify goals of anywidget project

manzt committed:
Add JS Cookiecutter deprecation notice

manzt committed:
Update widget graph figure caption

manzt committed:
Add limitations

manzt committed:
Update widgetstats.png

manzt committed:
Clarify api narrowing

nvictus commented:
Hi @mwcraig!

> I thought it would be helpful to preface the review by saying I'm a member of the Jupyter Widgets Council; I think I am able to do a fair review of the paper and that this doesn't pose a conflict of interest. I'm stating it up front, though, in case anyone involved feels otherwise. If so, I'm happy to step back as a reviewer.

Thank you for your review. We appreciate your fair and well-expressed feedback. As a member of the Jupyter Widgets Council, we think you are indeed well qualified to critique this paper.

> 1. The paper is very clear and well written. I have a couple of minor comments on the text/figures as inline comments.

Thank you for detailed feedback, we have addressed each individually in the comments.

> 2. How should the citations appear in the render paper? When I read the preview most of the citations should up as gold-ish text that didn't link to anything. Not sure if that is expected or not.

There was an error with the formatting due to an added citation which was incorrectly formatted. Thanks to @fwkoch for fixing this!

> 3. I think it would be useful to, at some point early in the paper, specify that the challenging part of writing a custom widget applies only to widgets that have front-end components. Widget packages that are python-only do not have that restriction, and can be useful.

This is a fair point. We are happy to make this qualification. However, to make sure we understand, when you say “python-only widgets”, we take it that you are referring to widgets that manipulate or compose other widgets which have front-end components (like Boxes). We have added a “Note” admonition to the manuscript to make this clear to the reader. We have also added a short paragraph to the manuscript to highlight that anywidgets derive from ipywidgets and therefore are interoperable with them.

> 4. I think there needs to be more detail about how you are calculating the number of widget packages that are ipywidgets vs anywidget vs converted.

We appreciate your thorough feedback and the need for more clarity. We've made the following improvements:

1. Updated the [README](https://github.com/manzt/anywidget-usage) with a detailed description of our semi-automated repository discovery process.
2. Added a statement to the README welcoming contributions and acknowledging limitations.
3. Revised the figure caption to include a high-level explanation of our approach.

Our process for identifying widget packages is as follows:
1. A nightly GitHub CI run executes a script (search.ts) that uses the GitHub code search API to find occurrences of `DOMWidgetModel` (for ipywidgets) and `anywidget.AnyWidget` (for anywidget) across GitHub.
2. New results are flagged for manual validation (https://github.com/manzt/anywidget-usage/issues/4) 
3. Trevor personally reviews each flagged repository, inspecting the files where the terms were found.
4. Verified repos are added using a command-line tool that creates an entry and updates the ignore list to prevent duplicate counting.

We acknowledge this process is imperfect and may miss some repositories, particularly older ipywidget-based projects with TypeScript or JavaScript components. Thank you for providing examples of missed projects – they have been reviewed and we improved our ipywidgets GitHub search API query and added many more ipywidgets repos since the original submission (https://github.com/manzt/anywidget-usage/pull/6).

We're adding a disclaimer about the limitations of GitHub API indexing and potential bias towards newer projects. However, we believe the overall trend and conclusion about the new energy in the widget ecosystem remains valid.

The repository is active and welcomes contributions. We're committed to improving our dataset and capturing the full scope of the widgets ecosystem. We invite the community to help identify missed projects and contribute to this effort.

> 5. The previous point doesn't take away from your main conclusion that the anywidget approach is easier than the cookie-cutter based approach -- it is definitely less complicated 😀! I also don't doubt that anywidget has increased the number of custom widgets being written that contain javascript.

Thank you for acknowledging this point. :)

> 6. I think there should be some acknowledgement of any downsides of the anywidget approach. The main one I see in my limited experience with anywidget is that importing javascript modules from URLs at runtime means they cannot be used offline.

Thanks for the comment. It's true that most of the examples and tutorials show using URLs because it's easy for prototyping. However, most "production" anywidgets use a bundler and work offline. Your module just needs to be ESM.

The main difference is that with traditional widgets, a bundle and multiple build systems are required. With anywidget, a single build system is optional but more straightforward because it just needs to target ESM. We have a project template CLI that is simpler than the cookiecutter and sets up projects with different frameworks to get you started. The templates we provide set up a bundler to bundle the code, but from the anywidget perspective, this is an option to add that complexity to the project if desired.

The largest downside at the moment is that the entry must be a single ESM file. We have added this limitation to the discussion.

manzt committed:
Callout ipywidgets comms protocol

manzt committed:
Comment on no-python scenarios

manzt commented:
> I am leaving an enthusiastic review of the paper, especially based on many of the conversations on integrating with MyST Markdown for scientific articles (like this one!). One of the things I would love to see better articulated is that widgets can work without the kernel (in many cases, e.g. educational widgets or static data set explorers).

Thanks for the review @rowanc1!

> This opens up a lot of possibilities for reuse of these modular widgets beyond jupyter compatible ecosystems, but are also compatible and can be upgraded by a kernel, when available. This opens possibilities when combined with tools like MyST, which may or may not have a kernel available.

Thanks for your enthusiasm and suggestion! We agree and have added a discussion section on anywidgets' use beyond Python and without a kernel. If you'd like stronger language regarding MyST, we're happy to make those changes.

manzt committed:
chore: Add jupyter-scatter reference

mwcraig commented:
Just wanted to let folks know I'll have some time to look this over/respond on Friday, Sep 13.
